lib/web.ts

```diff
@@ -119,80 +123,64 @@ export default class AutoConsent {
   }
 
   async _start() {
-    enableLogs && console.log(`Detecting CMPs on ${window.location.href}`)
+    enableLogs && console.log(`Detecting CMPs on ${window.location.href}`);
     this.updateState({ lifecycle: 'started' });
-    const cmps = await this.findCmp(this.config.detectRetries);
-    this.updateState({ detectedCmps: cmps.map(c => c.name)})
-    if (cmps.length > 0) {
-      const popupLookups: Promise<boolean>[] = [];
-      this.updateState({ lifecycle: 'cmpDetected' });
-      for (const cmp of cmps) {
-        enableLogs && console.log("detected CMP:", cmp.name, window.location.href);
-        this.sendContentMessage({
-          type: 'cmpDetected',
-          url: location.href,
-          cmp: cmp.name,
-        }); // notify the browser
-        popupLookups.push(this.waitForPopup(cmp).then((isOpen) => {
-          if (isOpen) {
-            if (!this.foundCmp) {
-              this.foundCmp = cmp;
-            }
-            this.updateState({ detectedPopups: this.state.detectedPopups.concat([cmp.name]) });
-            this.sendContentMessage({
-              type: 'popupFound',
-              cmp: cmp.name,
-              url: location.href,
-            }); // notify the browser
-            return true;
-          } else {
-            return Promise.reject(`${cmp.name} popup not found`);
-          }
-        }));
+    const { normal, cosmetic } = await this.findCmp(this.config.detectRetries);
+    this.updateState({ detectedCmps: [...normal.map(c => c.name), ...cosmetic.map(c => c.name)]})
+    if (normal.length === 0 && cosmetic.length === 0) {
+      enableLogs && console.log("no CMP found", location.href);
+      if (this.config.enablePrehide) {
+        this.undoPrehide();
       }
+      this.updateState({ lifecycle: 'nothingDetected' });
+      return false;
+    }
+    this.updateState({ lifecycle: 'cmpDetected' });
+    let foundPopups: AutoCMP[] = [];
+    if (normal.length > 0) {
+      foundPopups = await this.detectPopups(normal);
+    }
+    if (foundPopups.length === 0 && cosmetic.length > 0) {
+      foundPopups = await this.detectPopups(cosmetic);
+    }
```

Comment:
I'm not sure why we're handling these cases separately. Can't we just get a list of detected CMPs of both types, then just filter out `cosmetic` CMPs if there are any normal ones found?

Reply:
I meant more that I thought it would be cleaner if `findCmps` returned a single list of CMPs of both types, then we do the special case handing here. e.g.:
```javascript
foundPopups = await this.detectPopups(cmps.filter(r => !r.isCosmetic))
if (foundPopups.length === 0) {
  foundPopups = await this.detectPopups(cmps.filter(r => r.isCosmetic))
}
```
I think this avoids all the other cases where we have to check both `normal` and `cosmetic` arrays.

Reply:
Ah, I see. Yeah, that looks better, agreed